salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,835 @@
+    ),
+    # }}}
+    # /apis/rbac.authorization.k8s.io/v1beta1/ {{{
+    ('rbac.authorization.k8s.io/v1beta1', 'ClusterRole'): KindInfo(
+        model=k8s_client.V1beta1ClusterRole,
+        api_cls=k8s_client.RbacAuthorizationV1beta1Api,
+        name='cluster_role',
+    ),
+    ('rbac.authorization.k8s.io/v1beta1', 'ClusterRoleBinding'): KindInfo(
+        model=k8s_client.V1beta1ClusterRoleBinding,
+        api_cls=k8s_client.RbacAuthorizationV1beta1Api,
+        name='cluster_role_binding',
+    ),
+    ('rbac.authorization.k8s.io/v1beta1', 'Role'): KindInfo(
+        model=k8s_client.V1beta1Role,
+        api_cls=k8s_client.RbacAuthorizationV1beta1Api,
+        name='namespaced_role',
+    ),
+    ('rbac.authorization.k8s.io/v1beta1', 'RoleBinding'): KindInfo(
+        model=k8s_client.V1beta1RoleBinding,
+        api_cls=k8s_client.RbacAuthorizationV1beta1Api,
+        name='namespaced_role_binding',
+    ),
+    # }}}
+    # /apis/storage.k8s.io/v1/ {{{
+    ('storage.k8s.io/v1', 'StorageClass'): KindInfo(
+        model=k8s_client.V1StorageClass,
+        api_cls=k8s_client.StorageV1Api,
+        name='storage_class',
+    ),
+    # }}}
+}
+
+
+# CustomResources cannot rely on statically declared models, which is why their
+# management is treated differently from "standard" objects.
+
+class CustomApiClient(ApiClient):
+    CRUD_METHODS = dict(ApiClient.CRUD_METHODS, retrieve='get')
+
+    def __init__(self, group, version, kind, plural, scope):
+        self._group = group
+        self._version = version
+        self._scope = ObjectScoped(scope)
+        self._kind = kind
+        self._plural = plural
+
+        super(CustomApiClient, self).__init__(
+            api_cls=k8s_apis.CustomObjectsApi,
+            name='{}_custom_object'.format(self.scope.value)
+        )
+
+    group = property(operator.attrgetter('_group'))
+    version = property(operator.attrgetter('_version'))
+    scope = property(operator.attrgetter('_scope'))
+    kind = property(operator.attrgetter('_kind'))
+    plural = property(operator.attrgetter('_plural'))
+
+    def _method(self, verb):
+        """Return a CRUD method for this CustomApiClient.
+
+        This is constructed as a partial application of the appropriate
+        method from the `CustomObjectsApi`, and casts the resulting dict as
+        a `CustomObject`.
+        """
+        base_method = super(CustomApiClient, self)._method(verb)
+
+        def method(*args, **kwargs):
+            kwargs.update({
+                'group': self.group,
+                'version': self.version,
+                'plural': self.plural,
+            })
+
+            # Convert body to_dict if it's a CustomObject as
+            # `python-kubernetes` want a dict or a specific objects with
+            # some attributes like `swagger_types`, `attributes_map`, ...
+            if isinstance(kwargs.get('body'), CustomObject):
+                kwargs['body'] = kwargs['body'].to_dict()
```

Comment:
Why would this be the responsibility of the client? I'd argue the caller should take care of passing arguments in the correct format. If you want to catch errors early, maybe just:

```python
assert isinstance(kwargs.get('body', {}), dict), (
    "`body` must be a dict"
)
```

Reply:
If we do it client side then we can basically get rid of `CustomObjects` as we only want to do this for `CustomObjects` and not for "basic python-kubernetes objects".

Btw here we are able to make no real distinction between `CustomObjects` and "basic python-kubernetes objects" in the module and it's better IMO.